src/JsonSchema/SchemaFactory.php

```diff
@@ -230,14 +233,19 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
                 break;
             }
 
-            if ($type->isNullable()) {
-                $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
-            } else {
-                $propertySchema['$ref'] = $subSchema['$ref'];
-            }
+            $refs[] = ['$ref' => $subSchema['$ref']];
+            $isNullable = $isNullable ?? $type->isNullable();
+        }
+
+        if ($isNullable) {
+            $refs[] = ['type' => 'null'];
+        }
 
+        if (\count($refs) > 1) {
+            $propertySchema['anyOf'] = $refs;
             unset($propertySchema['type']);
-            break;
+        } elseif (1 === \count($refs)) {
+            $propertySchema = $refs[0];
```

Comment:
This erases other options in `$propertySchema`, I'm not sure it's safe. We should just add/override `$propertySchema['$ref']` option.